.github/workflows/canister-tests.yml

```diff
@@ -195,26 +195,120 @@ jobs:
           # Pull a deep copy so that we can set meaningful timestamps across builds, based on the commit history
           fetch-depth: '0'
 
+      # Figure out which commit modified what. Each commit defined here is used as a cache layer.
+      # We use those commit shas as cache keys. Since cargo uses timestamp-based comparison,
+      # each cache layer is retrieved and timestamped when the particular commit is checked out,
+      # so that checking out a later commit tells cargo which files are newer (since git will
+      # update the mtime of newer files)
+      - name: Retrieve commit information
+        id: git_info
+        run: |
+          # Last time dependencies or toolchain were updated; used for ~/cargo and dependencies
+          commit_deps=$(git log -1 --pretty=format:%H -- Cargo.* src/*/Cargo.toml rust-toolchain.toml)
+          echo "::set-output name=commit_deps::$commit_deps"
+          echo "deps: $commit_deps"
+
+          # Last time the canister code was modified
+          commit_canister=$(git log -1 --pretty=format:%H -- src/[^frontend]*)
```

Comment:
I think `[...]` makes it character based? At least in my quick test, this command included commits that touched stuff in `src/frontend` _only_.

Shouldn't this be `git log  -- . 'src/**' ':^src/frontend/**'`?

Also, if you really only want canister changes you could also additionally exclude `canister_tests` and `src/*/tests/**` (integration tests), i.e. `git log  -- . 'src/**' ':^src/frontend/**' ':^src/canister_tests/**' ':^src/*/tests/**'`

Reply:
> I think [...] makes it character based?

You're right! Looks like globs are a little more like regexes than I thought.

> Shouldn't this be `git log -- . 'src/**' ':^src/frontend/**'`?

Looks like this works, though I don't quite understand what it does... also I didn't know git itself supports globbing! What does `:` mean here?

Reply:
Its part of the git `pathspec` syntax: https://git-scm.com/docs/gitglossary#Documentation/gitglossary.txt-exclude